ddl/column_type_change_test.go

```diff
@@ -444,3 +444,225 @@ func (s *testColumnTypeChangeSuite) TestColumnTypeChangeFromIntegerToOthers(c *C
+	tk.MustGetErrCode("alter table t modify s timestamp", mysql.ErrUnsupportedDDLOperation)
+	tk.MustQuery("select * from t").Check(testkit.Rows("2020-07-15 18:32:18 2020-07-15 18:32:18 2020-07-15 18:32:18 2020-07-15 18:32:18 2020-07-15 18:32:18 2020-07-15 18:32:18 2020-07-15 18:32:17.888 2020-07-15 18:32:17.888"))
+	// year
+	reset(tk)
+	tk.MustExec("insert into t values ('2020', '91', '2', '2020', '20', '99', '2020-07-15 18:32:17.888', '2020-07-15 18:32:17.888')")
+	tk.MustExec("alter table t modify c year")
+	tk.MustExec("alter table t modify vc year")
+	tk.MustExec("alter table t modify bny year")
+	tk.MustExec("alter table t modify vbny year")
+	tk.MustExec("alter table t modify bb year")
+	tk.MustExec("alter table t modify txt year")
+	tk.MustExec("alter table t modify e year")
+	tk.MustExec("alter table t modify s year")
+	tk.MustQuery("select * from t").Check(testkit.Rows("2020 1991 2002 2020 2020 1999 2002 2002"))
+
+	// to json data types
+	reset(tk)
+	tk.MustExec("alter table t modify c char(15)")
+	tk.MustExec("alter table t modify vc varchar(15)")
+	tk.MustExec("alter table t modify bny binary(15)")
+	tk.MustExec("alter table t modify vbny varbinary(15)")
+	tk.MustExec("insert into t values ('{\"k1\": \"value\"}', '{\"k1\": \"value\"}', '{\"k1\": \"value\"}', '{\"k1\": \"value\"}', '{\"k1\": \"value\"}', '{\"k1\": \"value\"}', '{\"k1\": \"value\"}', '{\"k1\": \"value\"}')")
+	tk.MustExec("alter table t modify c json")
+	tk.MustExec("alter table t modify vc json")
+	tk.MustExec("alter table t modify bny json")
+	tk.MustExec("alter table t modify vbny json")
+	tk.MustExec("alter table t modify bb json")
+	tk.MustExec("alter table t modify txt json")
+	tk.MustExec("alter table t modify e json")
+	tk.MustExec("alter table t modify s json")
+	tk.MustQuery("select * from t").Check(testkit.Rows("{\"k1\": \"value\"} {\"k1\": \"value\"} {\"k1\": \"value\"} {\"k1\": \"value\"} {\"k1\": \"value\"} {\"k1\": \"value\"} \"{\\\"k1\\\": \\\"value\\\"}\" \"{\\\"k1\\\": \\\"value\\\"}\""))
+}
+
+func (s *testColumnTypeChangeSuite) TestColumnTypeChangeCastFailed(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	// Enable column change variable.
+	tk.Se.GetSessionVars().EnableChangeColumnType = true
+
+	// Set time zone to UTC
+	originalTz := tk.Se.GetSessionVars().TimeZone
+	tk.Se.GetSessionVars().TimeZone = time.UTC
+	defer func() {
+		tk.Se.GetSessionVars().EnableChangeColumnType = false
+		tk.Se.GetSessionVars().TimeZone = originalTz
+	}()
+
+	reset := func(tk *testkit.TestKit) {
+		tk.MustExec("drop table if exists t")
+		tk.MustExec("create table t (vc varchar(20))")
+	}
+
+	// wrong json data
+	reset(tk)
+	tk.MustExec("insert into t values ('{\"k1\": \"value\"')")
+	tk.MustGetErrCode("alter table t modify vc json", mysql.ErrInvalidJSONText)
+	// decimal bad number
+	reset(tk)
+	tk.MustExec("insert into t values ('abc')")
+	tk.MustGetErrCode("alter table t modify vc decimal(5,3)", mysql.ErrBadNumber)
```

Comment:
there are some cases be treated as an error while as a warning in TiDB.
for example:
"1a" will be truncated as 1 with waring in TiDB, but it errors in MySQL, this kind of incompatibility is acceptable, and should list in the test here.

Reply:
sure, I'll add tests for them.